sw/device/lib/crypto/impl/ecc/ecdh_p384.c

```diff
@@ -0,0 +1,164 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/impl/ecc/ecdh_p384.h"
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/hardened_memory.h"
+#include "sw/device/lib/crypto/drivers/otbn.h"
+#include "sw/device/lib/crypto/impl/ecc/p384_curve_point_valid.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+// Module ID for status codes.
+#define MODULE_ID MAKE_MODULE_ID('p', '3', 'x')
+
+OTBN_DECLARE_APP_SYMBOLS(p384_ecdh);          // The OTBN ECDSH/P-384 app.
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh, mode);    // ECDH application mode.
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh, dptr_x);  // The public key x-coordinate.
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh, dptr_y);  // The public key y-coordinate.
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh,
+                         x);  // The pointer to public key x-coordinate.
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh,
+                         y);  // The pointer to public key y-coordinate.
+OTBN_DECLARE_SYMBOL_ADDR(
+    p384_ecdh,
+    dptr_d0);  // The pointer to private key scalar d (share 0).
+OTBN_DECLARE_SYMBOL_ADDR(
+    p384_ecdh,
+    dptr_d1);  // The pointer to private key scalar d (share 1).
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh,
+                         d0);  // The private key scalar d (share 0).
+OTBN_DECLARE_SYMBOL_ADDR(p384_ecdh,
+                         d1);  // The private key scalar d (share 1).
+
+static const otbn_app_t kOtbnAppEcdh = OTBN_APP_T_INIT(p384_ecdh);
+static const otbn_addr_t kOtbnVarEcdhMode = OTBN_ADDR_T_INIT(p384_ecdh, mode);
+static const otbn_addr_t kOtbnVarEcdhDptrX =
+    OTBN_ADDR_T_INIT(p384_ecdh, dptr_x);
+static const otbn_addr_t kOtbnVarEcdhDptrY =
+    OTBN_ADDR_T_INIT(p384_ecdh, dptr_y);
+static const otbn_addr_t kOtbnVarEcdhX = OTBN_ADDR_T_INIT(p384_ecdh, x);
+static const otbn_addr_t kOtbnVarEcdhY = OTBN_ADDR_T_INIT(p384_ecdh, y);
+static const otbn_addr_t kOtbnVarEcdhDptrD0 =
+    OTBN_ADDR_T_INIT(p384_ecdh, dptr_d0);
+static const otbn_addr_t kOtbnVarEcdhDptrD1 =
+    OTBN_ADDR_T_INIT(p384_ecdh, dptr_d1);
+static const otbn_addr_t kOtbnVarEcdhD0 = OTBN_ADDR_T_INIT(p384_ecdh, d0);
+static const otbn_addr_t kOtbnVarEcdhD1 = OTBN_ADDR_T_INIT(p384_ecdh, d1);
+
+// Mode is represented by a single word. See `p384_ecdh.s` for values.
+static const uint32_t kOtbnEcdhModeWords = 1;
+static const uint32_t kOtbnEcdhModeKeypairRandom = 0x3f1;
+static const uint32_t kOtbnEcdhModeSharedKey = 0x5ec;
+// TODO static const uint32_t kOtbnEcdhModeKeypairFromSeed = 0x29f;
+// TODO static const uint32_t kOtbnEcdhModeSharedKeyFromSeed = 0x74b;
```

Comment:
nit: maybe these can go in an anonymous enum to match the other files?

Reply:
Good point. But then we should also change this in `ecdh_p256.c`. Because that's where I copied it from.